File: java/vortex-spark/common/src/main/scala/dev/vortex/spark/read/VortexScanBuilder.scala
// SPDX-License-Identifier: Apache-2.0
// SPDX-FileCopyrightText: Copyright the Vortex contributors

package dev.vortex.spark.read

import org.apache.spark.sql.SparkSession
import org.apache.spark.sql.connector.expressions.aggregate.{
  Aggregation,
  CountStar
}
import org.apache.spark.sql.connector.read.SupportsPushDownAggregates
import org.apache.spark.sql.execution.datasources.{
  AggregatePushDownUtils,
  PartitioningAwareFileIndex
}
import org.apache.spark.sql.execution.datasources.v2.FileScanBuilder
import org.apache.spark.sql.sources.Filter
import org.apache.spark.sql.types.StructType

import dev.vortex.spark.VortexOptions

object VortexScanBuilder {

  /** Option turning COUNT(*) pushdown off. */
  val AggregatePushdownOption = "vortex.aggregatePushdown"
}

/** Builds Vortex file scans with projection, filter, and COUNT(*) pushdown. */
final class VortexScanBuilder(
    sparkSession: SparkSession,
    fileIndex: PartitioningAwareFileIndex,
    schema: StructType,
    dataSchema: StructType,
    options: VortexOptions
) extends FileScanBuilder(sparkSession, fileIndex, dataSchema)
    with SupportsPushDownAggregates {

  private var finalSchema = new StructType()
  private var pushedAggregation = Option.empty[Aggregation]

  override protected def pushDataFilters(
      filters: Array[Filter]
  ): Array[Filter] =
    filters.filter(SparkFilterToVortexExpression.isPushable(_, dataSchema))

  override def pushAggregation(aggregation: Aggregation): Boolean = {
    if (!aggregatePushdownEnabled) {
      return false
    }
    pushedAggregationSchema(aggregation) match {
      case Some(aggregateSchema) =>
        finalSchema = aggregateSchema
        pushedAggregation = Some(aggregation)
        true
      case None => false
    }
  }

  // The reader emits one partial aggregate per file. Spark must combine files that share a group.
  override def supportCompletePushDown(aggregation: Aggregation): Boolean =
    false

  override def build(): VortexScan = {
    if (pushedAggregation.isEmpty) {
      finalSchema = readDataSchema()
    }
    new VortexScan(
      sparkSession,
      fileIndex,
      dataSchema,
      finalSchema,
      readPartitionSchema(),
      options,
      pushedDataFilters,
      pushedAggregation,
      partitionFilters,
      dataFilters
    )
  }

  /** Escape hatch matching `spark.sql.parquet.aggregatePushdown`, for when a
    * footer count must not be trusted.
    */
  private def aggregatePushdownEnabled: Boolean =
    options.getBoolean(VortexScanBuilder.AggregatePushdownOption, true)

  private def pushedAggregationSchema(
      aggregation: Aggregation
  ): Option[StructType] = {
    if (
      pushedDataFilters.nonEmpty ||
      !aggregation.aggregateExpressions().forall(_.isInstanceOf[CountStar])
    ) {
      None
    } else {
      AggregatePushDownUtils.getSchemaForPushedAggregation(
        aggregation,
        schema,
        partitionNameSet,
        dataFilters
      )
    }
  }
}

File: java/vortex-spark/src/main/java/dev/vortex/spark/read/VortexScanBuilder.java
// SPDX-License-Identifier: Apache-2.0
// SPDX-FileCopyrightText: Copyright the Vortex contributors

package dev.vortex.spark.read;

import static com.google.common.base.Preconditions.checkState;

import com.google.common.collect.ImmutableList;
import com.google.common.collect.Maps;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;
import org.apache.spark.sql.connector.catalog.CatalogV2Util;
import org.apache.spark.sql.connector.catalog.Column;
import org.apache.spark.sql.connector.expressions.NamedReference;
import org.apache.spark.sql.connector.expressions.Transform;
import org.apache.spark.sql.connector.expressions.aggregate.AggregateFunc;
import org.apache.spark.sql.connector.expressions.aggregate.Aggregation;
import org.apache.spark.sql.connector.expressions.aggregate.CountStar;
import org.apache.spark.sql.connector.expressions.filter.Predicate;
import org.apache.spark.sql.connector.read.Scan;
import org.apache.spark.sql.connector.read.ScanBuilder;
import org.apache.spark.sql.connector.read.SupportsPushDownAggregates;
import org.apache.spark.sql.connector.read.SupportsPushDownRequiredColumns;
import org.apache.spark.sql.connector.read.SupportsPushDownV2Filters;
import org.apache.spark.sql.types.DataType;
import org.apache.spark.sql.types.StructType;

/** Spark V2 {@link ScanBuilder} for table scans over Vortex files. */
public final class VortexScanBuilder
        implements ScanBuilder, SupportsPushDownRequiredColumns, SupportsPushDownV2Filters, SupportsPushDownAggregates {
    private final ImmutableList.Builder<String> paths;
    private final List<Column> tableColumns;
    private final List<Column> readColumns;
    private final Map<String, String> formatOptions;
    private final Set<String> partitionColumnNames;
    private Predicate[] pushedPredicates = new Predicate[0];
    private boolean pushedCountStar = false;

    /** Creates a new VortexScanBuilder with empty paths and columns. */
    public VortexScanBuilder(Map<String, String> formatOptions) {
        this(formatOptions, new Transform[0]);
    }

    /**
     * Creates a new VortexScanBuilder with empty paths and columns and the supplied partition transforms. Filters that
     * reference partition columns are not pushed down, since the partition columns are not stored inside the Vortex
     * files.
     */
    public VortexScanBuilder(Map<String, String> formatOptions, Transform[] partitionTransforms) {
        this.paths = ImmutableList.builder();
        Map<String, String> options = Maps.newHashMap();
        options.put("vortex.workerThreads", "4");
        options.putAll(formatOptions);
        this.tableColumns = new ArrayList<>();
        this.readColumns = new ArrayList<>();
        this.formatOptions = options;
        this.partitionColumnNames = collectPartitionColumnNames(partitionTransforms);
    }

    /**
     * Adds a file path to scan.
     *
     * @param path the file path to add
     * @return this builder for method chaining
     */
    public VortexScanBuilder addPath(String path) {
        this.paths.add(path);
        return this;
    }

    /**
     * Adds a column to read.
     *
     * @param column the column to add
     * @return this builder for method chaining
     */
    public VortexScanBuilder addColumn(Column column) {
        this.tableColumns.add(column);
        this.readColumns.add(column);
        return this;
    }

    /**
     * Adds multiple file paths to scan.
     *
     * @param paths the iterable of file paths to add
     * @return this builder for method chaining
     */
    public VortexScanBuilder addAllPaths(Iterable<String> paths) {
        this.paths.addAll(paths);
        return this;
    }

    /**
     * Adds multiple columns to read.
     *
     * @param columns the iterable of columns to add
     * @return this builder for method chaining
     */
    public VortexScanBuilder addAllColumns(Iterable<Column> columns) {
        for (Column column : columns) {
            addColumn(column);
        }
        return this;
    }

    /**
     * Builds a VortexScan with the configured paths and columns.
     *
     * <p>An empty column list is allowed: aggregates such as {@code count()} need no column data, and the scan then
     * reads the minimal schema.
     *
     * @return a new VortexScan instance
     * @throws IllegalStateException if no paths have been added
     */
    @Override
    public Scan build() {
        var paths = this.paths.build();

        checkState(!paths.isEmpty(), "paths cannot be empty");

        if (pushedCountStar) {
            return new VortexCountStarScan(paths, this.formatOptions);
        }

        return new VortexScan(
                paths,
                List.copyOf(this.tableColumns),
                List.copyOf(this.readColumns),
                pushedPredicates,
                this.formatOptions);
    }

    /**
     * Pushes down a global {@code COUNT(*)} so it is answered from file footer metadata instead of scanning data.
     *
     * <p>The pushdown is accepted only for an aggregation that is exactly one {@link CountStar} with no grouping
     * expressions, and only when no predicates were pushed down (a pushed filter changes the number of matching rows,
     * so footer row counts would over-count). Partition-column filters are never pushed (they are evaluated against
     * directory paths before planning), so they do not block the pushdown.
     *
     * <p>This is a partial pushdown ({@link #supportCompletePushDown(Aggregation)} stays {@code false}): the scan emits
     * one partial count per file and Spark performs the final summation.
     *
     * @param aggregation the aggregation Spark wants to push down
     * @return true if the aggregation was pushed, false to fall back to a regular scan
     */
    @Override
    public boolean pushAggregation(Aggregation aggregation) {
        if (pushedPredicates.length != 0) {
            return false;
        }
        if (aggregation.groupByExpressions().length != 0) {
            return false;
        }
        AggregateFunc[] aggregateFunctions = aggregation.aggregateExpressions();
        if (aggregateFunctions.length != 1 || !(aggregateFunctions[0] instanceof CountStar)) {
            return false;
        }
        this.pushedCountStar = true;
        return true;
    }

    /**
     * Prunes the columns to only include those specified in the required schema.
     *
     * <p>This method clears the current column list and replaces it with columns derived from the required schema.
     * Currently only supports top-level schema pruning - deeply nested schema pruning is not yet implemented.
     *
     * <p>After a successful {@link #pushAggregation(Aggregation)} this method is a no-op: Spark re-prunes to the
     * aggregation output schema, whose fields do not correspond to any table column.
     *
     * @param requiredSchema the schema specifying which columns are required
     */
    @Override
    public void pruneColumns(StructType requiredSchema) {
        if (pushedCountStar) {
            return;
        }
        readColumns.clear();
        readColumns.addAll(Arrays.asList(CatalogV2Util.structTypeToV2Columns(requiredSchema)));
    }

    /**
     * Splits the supplied predicates into pushed and not-pushed sets.
     *
     * <p>A predicate is pushed when it references only data columns (not partition columns) and uses operators and
     * literal types that {@link SparkPredicateToVortexExpression} can map to Vortex expressions. Predicates that
     * reference partition columns or use unsupported features are returned to Spark for post-scan evaluation.
     *
     * @return the predicates that Spark must still evaluate
     */
    @Override
    public Predicate[] pushPredicates(Predicate[] predicates) {
        Map<String, DataType> dataColumnTypes = new HashMap<>();
        for (Column column : readColumns) {
            if (!partitionColumnNames.contains(column.name())) {
                dataColumnTypes.put(column.name(), column.dataType());
            }
        }
        List<Predicate> pushed = new ArrayList<>();
        List<Predicate> postScan = new ArrayList<>();
        for (Predicate predicate : predicates) {
            if (SparkPredicateToVortexExpression.isPushable(predicate, dataColumnTypes)) {
                pushed.add(predicate);
            } else {
                postScan.add(predicate);
            }
        }
        this.pushedPredicates = pushed.toArray(new Predicate[0]);
        return postScan.toArray(new Predicate[0]);
    }

    /** Returns the predicates this scan promises to apply. */
    @Override
    public Predicate[] pushedPredicates() {
        return Arrays.copyOf(pushedPredicates, pushedPredicates.length);
    }

    private static Set<String> collectPartitionColumnNames(Transform[] transforms) {
        if (transforms == null || transforms.length == 0) {
            return Collections.emptySet();
        }
        Set<String> names = new HashSet<>();
        for (Transform transform : transforms) {
            for (NamedReference ref : transform.references()) {
                String[] parts = ref.fieldNames();
                if (parts.length == 1) {
                    names.add(parts[0]);
                }
            }
        }
        return names;
    }
}
